### PR TITLE
Export users command

### PR DESCRIPTION
**Pull request by rtibbles**
### Summary
* Adds `exportusers` management command
* Has optional `-d/--demographic-data` flag to toggle exporting of demographic data, defaults to `False`
* Updates `importusers` command to be able to import data directly from an edited form of the `exportusers` command

### Reviewer guidance
1) Export some user data
2) Update some demographic information
3) Import the csv using import users
4) Profit

Repeat and delete entire columns that are optional - like the facility id (as long as you have the name), and any other user information except the username.

Uses human readable column names, but does not currently implement any internationalization.
This could be extended to do so using the default device language (or potentially adding a language flag to the commands).

----

### Contributor Checklist

PR process:

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

Testing:

- [x] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [x] Critical and brittle code paths are covered by unit tests

### Reviewer Checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

File: kolibri/core/auth/csv_utils.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    mapped_obj = {}
+    for header, label in labels.items():
+        if header in mappings and header in obj:
+            mapped_obj[label] = mappings[header](obj)
+        elif header in obj:
+            mapped_obj[label] = obj[header]
+    return mapped_obj
+
+
+db_columns = (
```
Comment (jonboiser) [inline kolibri/core/auth/csv_utils.py:161]: Does this set the order? When I export, the headers come out like this (Demographic data is inserted between full name and username). My preference would be that the demo data come as a group of columns to the right of the basic model fields (facility to password)

```
Facility name,Facility id,Class name,Class id,Full name,Gender,Birth year,ID number,Username,Password
```

Comment (jonboiser) [reply kolibri/core/auth/csv_utils.py:161]: Also, I think the "password" field isn't being written. e.g., here's the first row of my export (data was generated using the UN Women CM executable during a hack session).

```
Facility name,Facility id,Class name,Class id,Full name,Gender,Birth year,ID number,Username,Password
a,6f963c47fefcfbe2d1aaec8490dba5da,,,j o r d a n,,,,jordan,
```

Comment (rtibbles) [reply kolibri/core/auth/csv_utils.py:161]: The `labels` OrderedDict defines the order, so can shuffle that to match.

The password can't be exported because we only store the hashed value. The only reason for including it here is to have the column header available in case someone wants to change the password.

File: kolibri/core/auth/management/commands/importusers.py [hunk 2]
```diff
@@ -24,90 +28,72 @@ def validate_username(user):
...
-    classroom = infer_and_create_class(user, facility)
+    facility = infer_facility(user.get("facility", None), facility=default_facility)
+    classroom = infer_and_create_class(user.get("class", None), facility)
     username = user["username"]
     try:
         user_obj = FacilityUser.objects.get(username=username, facility=facility)
         update_user_demographics(user, user_obj)
 
-        if classroom:
+        if classroom and not user_obj.is_member_of(classroom):
```
Comment (jonboiser) [inline kolibri/core/auth/management/commands/importusers.py:110]: I had one user in the export already in a classroom, but they fell through into this branch and got re-added into classroom. ~But there was another user (different classroom) that didn't go into this branch.~ (Actually the same user but different row) Not sure what the difference is between the two though.

File: kolibri/core/auth/management/commands/importusers.py [hunk 3]
```diff
@@ -24,90 +28,72 @@ def validate_username(user):
...
             classroom.add_member(user_obj)
             logger.info(
                 'Exiting user "{username}" was added to a classroom "{classroom}"'.format(
```
Comment (jonboiser) [inline kolibri/core/auth/management/commands/importusers.py]: Also just noticed this typo `exiting` -> `existing`

File: kolibri/core/auth/management/commands/importusers.py (hunk 2 above)
Comment (jonboiser) [reply kolibri/core/auth/management/commands/importusers.py:110]: Or at least the export gave them a 'classroom_id' in the CSV. Maybe they weren't actually in the classroom, hence they went into this branch.

Comment (jonboiser) [reply kolibri/core/auth/management/commands/importusers.py:110]: Okay, the problem seems to be that the CSV export is giving me these rows. Notice they are all for user `jb` (I need to test if it wil cause problems if I edit one or not the others, give them different, data, etc.)

The thing I'm noticing is that the first row for class "jb/...38c2" is that I don't actually see that Classroom when I run the app. I'm not sure what that's referring to or if that's pollution from a previous install. The other two classes are actually in the facility, and "JB" is a learner in them.

```
Facility name,Facility id,Class name,Class id,Full name,Gender,Birth year,ID number,Username,Password
a,6f963c47fefcfbe2d1aaec8490dba5da,jb,439ab8832b5b11d071204fd8aa5e38c2,jb,,,,jb,
a,6f963c47fefcfbe2d1aaec8490dba5da,My Class,4fb98d1aef10ce4bb49f380a2f661c8b,jb,,,,jb,
a,6f963c47fefcfbe2d1aaec8490dba5da,jajajaa,cbc3d5e53eaf3ea6c9e57ba59b394516,jb,,,,jb,

```

Comment (rtibbles) [reply kolibri/core/auth/management/commands/importusers.py:110]: Ah, I failed to document this, but this was the easiest way to manage a learner in multiple classes (rather than dynamically adding additional headers for each class, and with no way of preferentially selecting which class should be the one displayed).

Possible that the unidentified class is actually a group, so I'll double check the logic there.

Comment (rtibbles) [reply kolibri/core/auth/management/commands/importusers.py:110]: Confirmed, I failed to do that filter.

Comment (jonboiser) [reply kolibri/core/auth/management/commands/importusers.py:110]: I see. Maybe we can be more opinionated and assume that if we have the `-d` flag, then the primary goal is to display/edit demographic data. In that case, we could collapse multi-classroom users into a single row and have cell value like "Multiple classrooms" for that column.

Comment (rtibbles) [reply kolibri/core/auth/management/commands/importusers.py:110]: Ok - will do this as standard, as I don't see much of a use case for having the multiple lines, except as a mechanism for showing how to put users in multiple classes at once.

**Comment by radinamatic** [comment]
Is the hard requirement for _Facility name_ really necessary? Could we make it that if none is specified in the CSV file, and there's only one on the device, users get imported there?

`deb` asset installed on Ubuntu 16.04 gives the following error when trying to import users with the `kolibri manage importusers` command, successfully used in the past with the same CSV file.

```
osboxes@osboxes:~$ cd Desktop/
osboxes@osboxes:~/Desktop$ sudo dpkg -i kolibri-buildkite-build-5953-12101-kolibri_0.13.0.dev0+git.20190925180238-0ubuntu1_all.deb 
[sudo] password for osboxes: 
Selecting previously unselected package kolibri.
(Reading database ... 334845 files and directories currently installed.)
Preparing to unpack kolibri-buildkite-build-5953-12101-kolibri_0.13.0.dev0+git.20190925180238-0ubuntu1_all.deb ...
Unpacking kolibri (0.13.0.dev0+git.20190925180238-0ubuntu1) ...
Setting up kolibri (0.13.0.dev0+git.20190925180238-0ubuntu1) ...
Synchronizing state of kolibri.service with SysV init with /lib/systemd/systemd-sysv-install...
Executing /lib/systemd/systemd-sysv-install enable kolibri
Processing triggers for systemd (229-4ubuntu21.22) ...
Processing triggers for ureadahead (0.100.0-19.1) ...
osboxes@osboxes:~/Desktop$ kolibri manage importusers many-users.csv 
INFO     Option RUN_MODE in section [Deployment] being overridden by environment variable KOLIBRI_RUN_MODE
INFO     Running Kolibri with the following settings: kolibri.deployment.default.settings.base
INFO     Invoking command importusers many-users.csv
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/kolibri/core/auth/csv_utils.py", line 24, in infer_facility
    facility = Facility.objects.get(pk=facility_id)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/query.py", line 374, in get
    num = len(clone)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/query.py", line 232, in __len__
    self._fetch_all()
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/query.py", line 1121, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/query.py", line 53, in __iter__
    results = compiler.execute_sql(chunked_fetch=self.chunked_fetch)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/sql/compiler.py", line 876, in execute_sql
    sql, params = self.as_sql()
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/sql/compiler.py", line 441, in as_sql
    where, w_params = self.compile(self.where) if self.where is not None else ("", [])
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/sql/compiler.py", line 373, in compile
    sql, params = node.as_sql(self, self.connection)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/sql/where.py", line 79, in as_sql
    sql, params = compiler.compile(child)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/sql/compiler.py", line 373, in compile
    sql, params = node.as_sql(self, self.connection)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/lookups.py", line 170, in as_sql
    rhs_sql, rhs_params = self.process_rhs(compiler, connection)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/lookups.py", line 103, in process_rhs
    return self.get_db_prep_lookup(value, connection)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/lookups.py", line 196, in get_db_prep_lookup
    [get_db_prep_value(value, connection, prepared=True)]
  File "/usr/lib/python3/dist-packages/kolibri/dist/morango/models/fields/uuids.py", line 38, in get_db_prep_value
    value = uuid.UUID(value)
  File "/usr/lib/python3.5/uuid.py", line 140, in __init__
    raise ValueError('badly formed hexadecimal UUID string')
ValueError: badly formed hexadecimal UUID string

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/kolibri/core/auth/csv_utils.py", line 27, in infer_facility
    facility = Facility.objects.get(name=facility_id)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/query.py", line 380, in get
    self.model._meta.object_name
kolibri.core.auth.models.DoesNotExist: Facility matching query does not exist.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/bin/kolibri", line 11, in <module>
    load_entry_point('kolibri==0.13.0.dev0+git.20190925180238', 'console_scripts', 'kolibri')()
  File "/usr/lib/python3/dist-packages/kolibri/dist/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/kolibri/dist/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3/dist-packages/kolibri/dist/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python3/dist-packages/kolibri/utils/cli.py", line 235, in invoke
    return super(KolibriDjangoCommand, self).invoke(ctx)
  File "/usr/lib/python3/dist-packages/kolibri/dist/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3/dist-packages/kolibri/dist/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/kolibri/dist/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/usr/lib/python3/dist-packages/kolibri/utils/cli.py", line 633, in manage
    execute_from_command_line(["kolibri manage"] + ctx.args)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/core/management/__init__.py", line 364, in execute_from_command_line
    utility.execute()
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/core/management/__init__.py", line 356, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/core/management/base.py", line 283, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/core/management/base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "/usr/lib/python3/dist-packages/kolibri/core/auth/management/commands/importusers.py", line 210, in handle
    create_user(map_input(row), default_facility=default_facility)
  File "/usr/lib/python3/dist-packages/kolibri/core/auth/management/commands/importusers.py", line 89, in create_user
    facility = infer_facility(user.get("facility", None), facility=default_facility)
  File "/usr/lib/python3/dist-packages/kolibri/core/auth/csv_utils.py", line 31, in infer_facility
    facility=facility_id
ValueError: Facility matching identifier  was not found
osboxes@osboxes:~/Desktop$ 
```


The structure of the CSV file is as follows:
```
full_name,username,password,facility,class
Marc G.,marcg,marcg,,Primera
Peter B.,peterb,peterb,,Primera
Neela R.,neelar,neelar,,Primera
...
Archie M.2,archiem2,archiem2,,Segunda
Archie M.3,archiem3,archiem3,,Segunda
Archie M.4,archiem4,archiem4,,Segunda
```

[many-users.zip](https://github.com/learningequality/kolibri/files/3654826/many-users.zip)

**Comment by rtibbles** [comment]
Ah, thanks for testing @radinamatic - I think this was bad falsey checking here.

**Comment by rtibbles** [comment]
OK, think I have addressed all comments - let me know if there is still anything left that is causing concern.

**Comment by rtibbles** [comment]
Bah, still one valid test failure.

**Comment by radinamatic** [comment]
Installed the new DEB asset, previous issue is solved, but got a bunch of errors re-importing the users after adding some demographic data :confused: 

Ubuntu16 04 (start)  Running  - Oracle VM VirtualBox_215

Does not seem to be the value in the ID field (I was trying non-numeric values), as even those without it fail to import...

[many-users-demografic.zip](https://github.com/learningequality/kolibri/files/3655505/many-users-demografic.zip)

**Comment by radinamatic** [comment]
Ah, and there is an actual error in the Terminal:

```ERROR    Tried to update demographic info for "robertr17", but invalid values were given for fields: ['gender']
ERROR    Tried to update demographic info for "robertr18", but invalid values were given for fields: ['gender']
ERROR    Tried to update demographic info for "robertr19", but invalid values were given for fields: ['gender']
ERROR    Tried to update demographic info for "robertr20", but invalid values were given for fields: ['gender']
ERROR    Tried to update demographic info for "robertr21", but invalid values were given for fields: ['gender']
Traceback (most recent call last):
  File "/usr/bin/kolibri", line 11, in <module>
    load_entry_point('kolibri==0.13.0.dev0+git.20190925230214', 'console_scripts', 'kolibri')()
  File "/usr/lib/python3/dist-packages/kolibri/dist/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/kolibri/dist/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3/dist-packages/kolibri/dist/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python3/dist-packages/kolibri/utils/cli.py", line 235, in invoke
    return super(KolibriDjangoCommand, self).invoke(ctx)
  File "/usr/lib/python3/dist-packages/kolibri/dist/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3/dist-packages/kolibri/dist/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/kolibri/dist/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/usr/lib/python3/dist-packages/kolibri/utils/cli.py", line 633, in manage
    execute_from_command_line(["kolibri manage"] + ctx.args)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/core/management/__init__.py", line 364, in execute_from_command_line
    utility.execute()
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/core/management/__init__.py", line 356, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/core/management/base.py", line 283, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/core/management/base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "/usr/lib/python3/dist-packages/kolibri/core/auth/management/commands/importusers.py", line 212, in handle
    for row in reader:
  File "/usr/lib/python3.5/csv.py", line 110, in __next__
    row = next(self.reader)
_csv.Error: line contains NULL byte
osboxes@osboxes:~/Desktop$ 
```

**Comment by rtibbles** [comment]
Guessing that this is some sort of encoding issue.

**Comment by radinamatic** [comment]
Deleting the incorrectly formatted entries made the import possible:

Ubuntu16 04 (start)  Running  - Oracle VM VirtualBox_219

Unless there is still some edge case you think we should test for @rtibbles, I'll grab the latest asset to make one last sanity check, and approve if all goes well!

**Comment by rtibbles** [comment]
Weird - let's keep an eye out for it in the future just in case.

**Comment by radinamatic** [comment]
Ok, this is from the latest asset - would it be possible to catch this error more gracefully, similar to the previous? Something akin to:

`CommandError: Facility name/id mismatch during the import process. Please make sure that the facility name/id in the CSV file are the same as on the device.`

This could also avoid having the Kolibri crash report appear :wink: 

Ubuntu16 04 (start)  Running  - Oracle VM VirtualBox_220

A part from that, all the subsequent import/export operations with editing users inside Kolibri, and editing CSV outside, went well :+1: 

image

**Comment by rtibbles** [comment]
Updated to raise a `CommandError` in this instance.

**Comment by radinamatic** [comment]
:pray: :bowing_woman: 

Ubuntu16 04 (start)  Running  - Oracle VM VirtualBox_229

**Review by radinamatic — APPROVED** [review]
LGTM! :+1:
